## Let's examine how many unique D/A pairs and unique molecules we have in the dataset

In [1]:
from rdkit import Chem
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

Open the dataset csv

In [2]:
df = pd.read_csv('../OSC_dataset_metaanalysis.csv')
df

NFA_ID reported_acceptor       NFA_other_names  don_ID reported_donor  \
0          0            BTz-Np                BTz-Np       0           P3HT   
1          1            NTz-Np                NTz-Np       0           P3HT   
2          2              ITIC  ITIC, IT-IC, ITIC-0F       0           P3HT   
3          2              ITIC  ITIC, IT-IC, ITIC-0F       0           P3HT   
4          2              ITIC  ITIC, IT-IC, ITIC-0F       1       PBDT-NQx   
...      ...               ...                   ...     ...            ...   
1970    1037          TT-Se-2F              TT-Se-2F      15            D18   
1971    1038             PY-DT                 PY-DT      12            PM6   
1972    1039               M80              M80, M36      12            PM6   
1973    1039               M36                   M36      12            PM6   
1974    1040               M81                   M81      12            PM6   

                                 don_other_names  \
0                                           P3HT   
1                                           P3HT   
2                                           P3HT   
3                                           P3HT   
4                                       PBDT-NQx   
...                                          ...   
1970                                         D18   
1971  PBDB-TF, PM6, PBDB-T-2F, PBDB-T2F, PBDBT-F   
1972  PBDB-TF, PM6, PBDB-T-2F, PBDB-T2F, PBDBT-F   
1973  PBDB-TF, PM6, PBDB-T-2F, PBDB-T2F, PBDBT-F   
1974  PBDB-TF, PM6, PBDB-T-2F, PBDB-T2F, PBDBT-F   

                               DOI    Voc    Jsc     FF  ...  PCE_std  \
0           10.1002/adfm.201504153  0.940   3.52  46.00  ...     0.02   
1           10.1002/adfm.201504153  0.900   5.18  60.00  ...     0.08   
2           10.1002/adma.201601803  0.520   4.22  56.93  ...      NaN   
3               10.1002/bkcs.12614  0.520   4.22  57.00  ...      NaN   
4           10.1002/adfm.201701491  0.870  16.21  64.60  ...      NaN   
...                            ...    ...    ...    ...  ...      ...   
1970    10.1016/j.xcrp.2022.101169  0.760  23.82  67.50  ...     0.16   
1971        10.1002/adma.202208165  0.957  23.15  75.70  ...     0.15   
1972  10.1016/j.nanoen.2022.108116  0.820   3.04  56.55  ...     0.10   
1973  10.1016/j.nanoen.2022.108116  0.910  23.93  78.40  ...     0.10   
1974  10.1016/j.nanoen.2022.108116  0.860  21.73  59.41  ...     0.10   

      acceptor_type  acceptor_class fused_rings_in_core donor_type  \
0                sm         unfused                 3.0    polymer   
1                sm           fused                 4.0    polymer   
2                sm           fused                 7.0    polymer   
3                sm           fused                 7.0    polymer   
4                sm           fused                 7.0    polymer   
...             ...             ...                 ...        ...   
1970             sm         unfused                 2.0    polymer   
1971        polymer             NaN                 NaN    polymer   
1972             sm           fused                 7.0    polymer   
1973             sm           fused                 7.0    polymer   
1974             sm           fused                 7.0    polymer   

                                  acc_SMILES_sidechains  \
0     O=C1C=2C=CC=C3C(C#CC=4SC(=CC4CCCCCC)C5=CC=C(C=...   
1     O=C1C=2C=CC=C3C(C#CC=4SC(=CC4CCCCCC)C5=CC=6C7=...   
2     N#CC(C#N)=C1C=2C=CC=CC2C(=O)C1=CC=3SC4=C(SC=5C...   
3     N#CC(C#N)=C1C=2C=CC=CC2C(=O)C1=CC=3SC4=C(SC=5C...   
4     N#CC(C#N)=C1C=2C=CC=CC2C(=O)C1=CC=3SC4=C(SC=5C...   
...                                                 ...   
1970  N#C/C(C#N)=C1C(C=C(C(F)=C2)F)=C2C(C/1=C\C([Se]...   
1971  O=C1C2=CC=C(C=C2C(/C1=C/C3=C(C4=C(S3)C(N(CC(CC...   
1972  O=C(C/1=C\C2=CC3=C(S2)C(OC4=C5C(OCC(CCCCCC)CCC...   
1973  O=C(C1=C(C/2=C(C#N)\C#N)C=C(F)C(F)=C1)C2=C/C3=...   
1974  O=C(C1=C(C/2=C(C#N)\C#N)C=C(F)C(F)=C1)C2=C/C3=...   

     

How many unique NFAs do we have (with and without side chains)?

In [3]:
unique_smiles = []

for x in range(len(df['acc_SMILES_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_sidechains'][x]))

        if canon_smi not in unique_smiles:
            unique_smiles.append(canon_smi)
    except:
        print(x)
        print(df['acc_SMILES_sidechains'][x])

print('The number of unique NFAs with side chains is: %s' % len(unique_smiles))

The number of unique NFAs with side chains is: 1082


In [4]:
unique_smiles = []

for x in range(len(df['acc_SMILES_shortened_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_shortened_sidechains'][x]))

        if canon_smi not in unique_smiles:
            unique_smiles.append(canon_smi)
    except:
        print(x)
        print(df['acc_SMILES_shortened_sidechains'][x])

print('The number of unique NFAs with shortened side chains is: %s' % len(unique_smiles))

The number of unique NFAs with shortened side chains is: 891


### There are 1082 unique NFAs, and only 891 when you shorten the alkyl side chains to ethyl groups

How many unique donors are there?

In [5]:
unique_smiles = []

for x in range(len(df['don_SMILES_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_sidechains'][x]))

        if canon_smi not in unique_smiles:
            unique_smiles.append(canon_smi)
    except:
        print(x)
        print(df['don_SMILES_sidechains'][x])

print('The number of unique donors with side chains is: %s' % len(unique_smiles))

The number of unique donors with side chains is: 432


In [6]:
unique_smiles = []

for x in range(len(df['don_SMILES_shortened_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_shortened_sidechains'][x]))

        if canon_smi not in unique_smiles:
            unique_smiles.append(canon_smi)
    except:
        print(x)
        print(df['don_SMILES_shortened_sidechains'][x])

print('The number of unique donors with shortened alkyl side chains is: %s' % len(unique_smiles))

The number of unique donors with shortened alkyl side chains is: 374


### There are 432 unique donors, and only 374 when you shorten the alkyl side chains to ethyl groups

How many unique pairs are there in the dataset?

In [7]:
print('There are %s total pairs in the OPEP3 dataset' %len(df))

There are 1975 total pairs in the OPEP3 dataset


In [8]:
unique_pairs = []
unique_smiles_NFA = []
unique_smiles_donors = []

for x in range(len(df['don_SMILES_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi_don = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_sidechains'][x]))
        canon_smi_acc = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_sidechains'][x]))

        canon_pair = [canon_smi_don, canon_smi_acc]

        if canon_pair not in unique_pairs:
            unique_pairs.append(canon_pair)
    except:
        print(x)

print('The number of unique pairs with full side chains is: %s' % len(unique_pairs))

The number of unique pairs with full side chains is: 1751


In [9]:
unique_pairs = []
unique_smiles_NFA = []
unique_smiles_donors = []

for x in range(len(df['don_SMILES_shortened_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi_don = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_shortened_sidechains'][x]))
        canon_smi_acc = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_shortened_sidechains'][x]))

        canon_pair = [canon_smi_don, canon_smi_acc]

        if canon_pair not in unique_pairs:
            unique_pairs.append(canon_pair)
    except:
        print(x)

print('The number of unique pairs with shortened alkyl side chains is: %s' % len(unique_pairs))

The number of unique pairs with shortened alkyl side chains is: 1541


### Lets save these as filtered dataset csvs. With sidechains means it was filtered by removing duplicates among SMILES that contain the sidechains. No sidechains means only the SMILES with shortened sidechains were looked at.

The csv was manually ordered by PCE, so highest PCE is first and lowest PCE is last. This is necessary so we can select the top performing pair for the filtered dataset

In [10]:
df_filtered = pd.DataFrame(columns=['reported_acceptor','reported_donor', 'reference', 'DOI', 'Voc', 'Jsc',	'FF', 'PCE', 'acceptor_type', 'acceptor_class', 'fused_rings_in_core', 'donor_type', 'acc_SMILES_sidechains',	'acc_SMILES_shortened_sidechains', 'don_SMILES_sidechains',	'don_SMILES_shortened_sidechains'])

unique_pairs = []

for x in range(len(df['don_SMILES_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi_don = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_sidechains'][x]))
        canon_smi_acc = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_sidechains'][x]))

        canon_pair = [canon_smi_don, canon_smi_acc]

        if canon_pair not in unique_pairs:
            unique_pairs.append(canon_pair)
            df_filtered.loc[x] = df.iloc[x]
    except:
        pass

df_filtered.to_csv('OSC_filtered_with_sidechains.csv')

In [11]:
unique_pairs = []
df_filtered_no_sidechains = pd.DataFrame(columns=['reported_acceptor','reported_donor', 'reference', 'DOI', 'Voc', 'Jsc',	'FF', 'PCE', 'acceptor_type', 'acceptor_class', 'fused_rings_in_core', 'donor_type', 'acc_SMILES_sidechains',	'acc_SMILES_shortened_sidechains', 'don_SMILES_sidechains',	'don_SMILES_shortened_sidechains'])


for x in range(len(df['don_SMILES_shortened_sidechains'])):
    try:
        # convert each to canonical smiles for fair comparison
        canon_smi_don = Chem.MolToSmiles(Chem.MolFromSmiles(df['don_SMILES_shortened_sidechains'][x]))
        canon_smi_acc = Chem.MolToSmiles(Chem.MolFromSmiles(df['acc_SMILES_shortened_sidechains'][x]))

        canon_pair = [canon_smi_don, canon_smi_acc]

        if canon_pair not in unique_pairs:
            unique_pairs.append(canon_pair)
            df_filtered_no_sidechains.loc[x] = df.iloc[x]
    except:
        print(x)

df_filtered_no_sidechains.to_csv('OSC_filtered_no_sidechains.csv')